c++/src/kj/source-location.h

```diff
@@ -0,0 +1,91 @@
+// Licensed under the MIT License:
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+#pragma once
+
+#include "string.h"
+
+#if __clang__ && __clang_major__ >= 9
+#define KJ_CALLER_COLUMN() __builtin_COLUMN()
+// GCC does not implement __builtin_COLUMN() as that's non-standard but MSVC & clang do.
+// MSVC does as of version https://github.com/microsoft/STL/issues/54) but there's currently not any
+// pressing need for this for MSVC & writing the write compiler version check is annoying.
+#else
+#define KJ_CALLER_COLUMN() 0
+#endif
+
+#if __cplusplus > 201703L
+#define KJ_COMPILER_SUPPORTS_SOURCE_LOCATION 1
+#elif __clang__ && __clang_major__ >= 9
+// Clang 9 added these builtins: https://releases.llvm.org/9.0.0/tools/clang/docs/LanguageExtensions.html
+#define KJ_COMPILER_SUPPORTS_SOURCE_LOCATION 1
+#elif __GNUC__ >= 5
+// GCC 5 supports the required builtins: https://gcc.gnu.org/onlinedocs/gcc-5.1.0/gcc/Other-Builtins.html
+#define KJ_COMPILER_SUPPORTS_SOURCE_LOCATION 1
+#endif
+
+namespace kj {
+class SourceLocation {
+  // libc++ doesn't seem to implement <source_location> (or even <experimental/source_location>), so
+  // this is a non-STL wrapper over the compiler primitives (these are the same across MSVC/clang/
+  // gcc). Additionally this uses kj::StringPtr for holding the strings instead of const char* which
+  // makes it integrate a little more nicely into KJ.
+public:
+#if KJ_COMPILER_SUPPORTS_SOURCE_LOCATION
+  static constexpr SourceLocation current(
+      const char* file = __builtin_FILE(), const char* function = __builtin_FUNCTION(),
+      uint line = __builtin_LINE(), uint column = KJ_CALLER_COLUMN()) {
+    // When invoked with no arguments returns a SourceLocation that points at the caller of the
+    // function.
+    return SourceLocation(file, function, line, column);
+  }
+#endif
+
+  constexpr SourceLocation() : SourceLocation("??", "??", 0, 0) {}
```

Comment:
Is there a good reason not to have SourceLocation's constructor default to initializing it to the line that called it? Then the separate `current()` call wouldn't be needed, and you could use `{}` as the default initializer which is much more concise.

I ask because I wrote some code yesterday that took advantage of __builtin_FILE() / __builtin_LINE() and I did it this way and thought it worked out pretty nicely.

Reply:
Not that I can think of. As I mentioned, I was just following the spec for `std::source_location` & I figured there must be some reason why it was a separate function. I'll change it to be automatic for the default constructor.

Reply:
I guess the one weird thing is that you can't have an "explicit" on the constructor. That means that if you pass in a string in the wrong spot you'll "corrupt" the SourceLocation with no warning from the compiler.

```
void foo(SourceLocation = {});

foo(); // OK - this is what we want.
foo("abc"); // BAD - no warning since the constructor can't be explicit or `{}` wouldn't work
```

I think that's the reason the C++ standard went with an explicit function as it avoids these problems. Let me know what you'd like (generally I'm paranoid about default arguments being eager to swallow accidents).

Reply:
Interesting point. It looks like the following construction is impossible to call with non-default arguments:

```
class SourceLocation {
private:
  struct Badge { explicit Badge() {} };
public:
  SourceLocation(Badge = Badge(), const char* file = __builtin_FILE(), uint line = __builtin_LINE())
      : file(file), line(line) {}
  ...
};
```

Reply:
Clever. Applied.